includes/admin/extensions/class-extension-manager.php

```diff
@@ -0,0 +1,524 @@
+		$required_pass = filter_input( INPUT_POST, 'pass', FILTER_SANITIZE_STRING );
+		if ( ! $type ) {
+			wp_send_json_error( $generic_error );
+		}
+
+		// Check if new installations are allowed.
+		if ( ! $this->can_install( $type, $required_pass ) ) {
+			wp_send_json_error( $generic_error );
+		}
+
+		$error = 'plugin' === $type
+			? esc_html__( 'Could not install the plugin. Please download and install it manually via Plugins > Add New.', 'easy-digital-downloads' )
+			: esc_html__( 'Could not install the extension. Please download it from edd.com and install it manually.', 'easy-digital-downloads' );
+
+		$plugin = filter_input( INPUT_POST, 'plugin', FILTER_SANITIZE_STRING );
+		if ( empty( $plugin ) ) {
+			wp_send_json_error( $error );
+		}
+
+		// Set the current screen to avoid undefined notices.
+		set_current_screen( 'edd_page_edd-settings' );
+
+		// Prepare variables.
+		$url = esc_url_raw(
+			add_query_arg(
+				array(
+					'post_type' => 'download',
+					'page'      => 'edd-addons',
+				),
+				admin_url( 'edit.php' )
+			)
+		);
+
+		ob_start();
+		$creds = request_filesystem_credentials( $url, '', false, false, null );
+
+		// Hide the filesystem credentials form.
+		ob_end_clean();
+
+		// Check for file system permissions.
+		if ( ! $creds ) {
+			wp_send_json_error( $error );
+		}
+
+		if ( ! WP_Filesystem( $creds ) ) {
+			wp_send_json_error( $error );
+		}
+
+		/*
+		 * We do not need any extra credentials if we have gotten this far, so let's install the plugin.
+		 */
+		require_once EDD_PLUGIN_DIR . 'includes/admin/installers/class-plugin-silent-upgrader.php';
+		require_once EDD_PLUGIN_DIR . 'includes/admin/installers/class-plugin-silent-upgrader-skin.php';
+		require_once EDD_PLUGIN_DIR . 'includes/admin/installers/class-install-skin.php';
+
+		// Do not allow WordPress to search/download translations, as this will break JS output.
+		remove_action( 'upgrader_process_complete', array( 'Language_Pack_Upgrader', 'async_upgrade' ), 20 );
+
+		// Create the plugin upgrader with our custom skin.
+		$installer = new \EDD\Admin\Installers\PluginSilentUpgrader( new \EDD\Admin\Installers\Install_Skin() );
+
+		// Error check.
+		if ( ! method_exists( $installer, 'install' ) || empty( $plugin ) ) {
+			wp_send_json_error( $error );
+		}
+
+		$installer->install( $plugin ); // phpcs:ignore
+
+		// Flush the cache and return the newly installed plugin basename.
+		wp_cache_flush();
+
+		$plugin_basename = $installer->plugin_info();
+
+		if ( empty( $plugin_basename ) ) {
+			wp_send_json_error( $error );
+		}
+
+		$result = array(
+			'msg'          => $generic_error,
+			'is_activated' => false,
+			'basename'     => $plugin_basename,
+		);
+
+		// Check for permissions.
+		if ( ! current_user_can( 'activate_plugins' ) ) {
+			$result['msg'] = 'plugin' === $type ? esc_html__( 'Plugin installed.', 'easy-digital-downloads' ) : esc_html__( 'Extension installed.', 'easy-digital-downloads' );
+
+			wp_send_json_success( $result );
```

Comment:
Need to account for this result--if a user can install, but not activate a plugin, what needs to happen?